File: frontend/src/types/message.ts
export enum AuthorType {
  USER = "user",
  AGENT = "agent",
}

export type MessageStatus = "pending" | "sent" | "delivered" | "failed"

// Structured data interfaces for message cards
export interface ArtifactData {
  artifact_id: string
  artifact_type: string
  title: string
  description?: string
  version: number
  status: string
  agent_name?: string
}

export interface StoriesCreatedData {
  count: number
  story_ids: string[]
  prd_artifact_id?: string
}

export interface AgentQuestionData {
  question_id: string
  question_text: string
  question_type: "open" | "multichoice"
  options?: string[]
  allow_multiple?: boolean
  context?: any
}

export interface MessageAttachment {
  type: "document"
  filename: string
  file_path: string
  file_size: number
  mime_type?: string
  extracted_text?: string
}

export type Message = {
  id: string
  project_id: string
  author_type: AuthorType
  user_id?: string
  agent_id?: string
  agent_name?: string // "Team Leader" | "Business Analyst" | "Developer" | "Tester"
  persona_avatar?: string // Agent's persona avatar URL
  content: string
  message_type?: string // "text" | "artifact_created" (PRD, analysis, etc.) | "stories_created" | "agent_question" | "document_upload" | etc.
  structured_data?: ArtifactData | StoriesCreatedData | AgentQuestionData | any // JSON data for cards/previews
  message_metadata?: any // Message metadata (agent_name, preview_id, incomplete_flag, etc.)
  attachments?: MessageAttachment[] // File attachments (for document uploads)
  created_at: string
  updated_at: string
  status?: MessageStatus // Message delivery status (for user messages)
}

export type MessagesPage = {
  data: Message[]
  count: number
}

File: backend/app/schemas/message.py
"""Message-related schemas."""

from datetime import datetime
from typing import Any, Optional
from uuid import UUID

from sqlmodel import SQLModel

from app.models import AuthorType


class ChatMessageBase(SQLModel):
    content: str
    author_type: AuthorType


class ChatMessageCreate(ChatMessageBase):
    project_id: UUID
    agent_id: Optional[UUID] = None
    message_type: Optional[str] = "text"
    structured_data: Optional[dict] = None
    attachments: Optional[list[dict]] = None


class ChatMessageUpdate(SQLModel):
    content: Optional[str] = None


class ChatMessagePublic(SQLModel):
    id: UUID
    project_id: UUID
    author_type: AuthorType
    user_id: Optional[UUID] = None
    agent_id: Optional[UUID] = None
    agent_name: Optional[str] = None  # Agent's human-readable name
    persona_avatar: Optional[str] = None  # Agent's persona avatar URL
    content: str
    message_type: Optional[str] = "text"
    structured_data: Optional[Any] = None
    message_metadata: Optional[dict] = None
    attachments: Optional[list[dict]] = None
    created_at: datetime
    updated_at: datetime


class ChatMessagesPublic(SQLModel):
    data: list[ChatMessagePublic]
    count: int
